pkg/infra/pulumi.go

```diff
@@ -0,0 +1,275 @@
+package infra
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+
+	"github.com/AudiusProject/audius-d/pkg/conf"
+	"github.com/AudiusProject/audius-d/pkg/logger"
+	"github.com/pulumi/pulumi-aws/sdk/v6/go/aws/ec2"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optdestroy"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optpreview"
+	"github.com/pulumi/pulumi/sdk/v3/go/auto/optup"
+	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+)
+
+var (
+	confCtxConfig *conf.ContextConfig
+)
+
+func init() {
+	var err error
+	confCtxConfig, err = conf.ReadOrCreateContextConfig()
+	if err != nil {
+		logger.Error("Failed to retrieve context. ", err)
+		return
+	}
+}
+
+func pulumiInit() {
+	baseDir, err := conf.GetConfigBaseDir()
+	if err != nil {
+		logger.Error("Failed to retrieve config base dir. ", err)
+		return
+	}
+
+	envVars := map[string]string{
+		// local pulumi stacks require this passphrase env var
+		// we are not using pulumi secrets - so this is not a security risk
+		"PULUMI_CONFIG_PASSPHRASE": "",
+		// use a single ~/.audius/.pulumi for all pulumi state management files
+		// the .pulumi dir is created upon pulumi login below
+		"PULUMI_HOME": fmt.Sprintf("%s/.pulumi", baseDir),
+	}
+	err = setMultipleEnvVars(envVars)
+	if err != nil {
+		logger.Error("Error setting environment variables: %v", err)
+		return
+	}
+	err = pulumiLogin(baseDir)
+	if err != nil {
+		if execErr, ok := err.(*exec.Error); ok && execErr.Err == exec.ErrNotFound {
+			confirmation, err := promptConfirmation(
+				"'audius-ctl infra' requires Pulumi (https://pulumi.io).\nPulumi is not installed. Would you like to install it now? [y/N]: ")
+			if err != nil {
+				logger.Error("Error reading confirmation:", err)
+				return
+			}
+			if confirmation {
+				if installErr := installPulumi(); installErr != nil {
+					logger.Error("Failed to install Pulumi: %v\n", installErr)
+				}
+				logger.Info("Pulumi installed successfully.")
+				if err := pulumiLogin(baseDir); err != nil {
+					logger.Error("Failed to login to Pulumi: %v\n", err)
+				}
+			} else {
+				logger.Info("Pulumi installation canceled.")
+			}
+			os.Exit(0)
```

Comment:
I was surprised that the command exited after installing pulumi, I expected it to just continue. If we move the confirmation and installation portion to cmd/audius-ctl/infra.go, then we won't have to exit here.